File: tools/shv-interface-codegen/tests/interface_test.py
from pathlib import Path
import argparse, copy, hashlib, importlib.util, json, shutil, subprocess, tempfile, unittest

H = Path(__file__).resolve().parents[1]
R = H.parents[1]
s = importlib.util.spec_from_file_location("generator", H / "generate.py")
g = importlib.util.module_from_spec(s)
s.loader.exec_module(g)
p = argparse.ArgumentParser()
p.add_argument("--owner", required=True)
p.add_argument("--engine", required=True)
a, rest = p.parse_known_args()
M = R / "modules" / a.owner


class InterfaceTests(unittest.TestCase):
    def test_frozen_native_parity(self):
        d = g.read(M / "OWNER-INTERFACE.json")
        hist = g.read(M / "tests/fixtures/interface-history.v1.json")["descriptors"]
        expected = copy.deepcopy(list(hist.values())[-1])
        expected["engine_version"] = d["current_version"]
        expected.pop("descriptor_digest")
        expected["descriptor_digest"] = (
            "sha256:"
            + hashlib.sha256(
                json.dumps(expected, sort_keys=True, separators=(",", ":")).encode()
            ).hexdigest()
        )
        self.assertEqual(
            json.loads(subprocess.check_output([a.engine, "--descriptor"])), expected
        )

    def test_deterministic_projection_and_drift(self):
        d = g.validate(g.read(M / "OWNER-INTERFACE.json"))
        self.assertEqual(g.outputs(d), g.outputs(copy.deepcopy(d)))
        g.emit(d, check=True)
        with tempfile.TemporaryDirectory() as td:
            root = Path(td)
            g.emit(d, root)
            path = next(iter(g.outputs(d)))
            (root / path).write_text("changed")
            with self.assertRaises(AssertionError):
                g.emit(d, root, check=True)

    def test_declaration_rejections(self):
        original = g.read(M / "OWNER-INTERFACE.json")
        old = list(original["releases"])[0]
        mutations = {
            "extra": lambda d: d.update(extra=True),
            "missing": lambda d: d.pop("schemas"),
            "owner": lambda d: d.update(module_id="unknown"),
            "namespace": lambda d: d.update(namespace="foreign"),
            "release": lambda d: d["releases"].update(latest=[]),
            "old_admission": lambda d: d["releases"][old].append("unknown"),
            "duplicate": lambda d: d["operations"].append(d["operations"][0]),
            "mutation": lambda d: d["operations"][0].update(mutability="write"),
            "protocol": lambda d: d["operations"][0].update(output_protocol="invented"),
            "interaction": lambda d: d["operations"][0].update(
                administrative_interactions=["mutate"]
            ),
            "reader": lambda d: d["embedded_dependencies"].append(
                {"engine_id": "foreign", "version": "latest"}
            ),
            "escape": lambda d: d["companions"].append("../escape"),
            "duplicate_schema": lambda d: d["schemas"].append(d["schemas"][0]),
            "missing_schema": lambda d: d["schemas"].pop(),
        }
        for name, mutate in mutations.items():
            with self.subTest(name=name):
                d = copy.deepcopy(original)
                mutate(d)
                with self.assertRaises((AssertionError, KeyError)):
                    g.validate(d)

    def test_duplicate_keys_and_schema_refs(self):
        with tempfile.TemporaryDirectory() as td:
            p = Path(td) / "dup.json"
            p.write_text('{"a":1,"a":2}')
            with self.assertRaises(ValueError):
                g.read(p)
        d = g.read(M / "OWNER-INTERFACE.json")
        for ref in ["../../outside.json", "#/$defs/Absent"]:
            with self.subTest(ref=ref), tempfile.TemporaryDirectory() as td:
                root = Path(td)
                for path in [
                    "modules/" + a.owner + "/tests/fixtures/interface-history.v1.json",
                    *d["schemas"],
                    *d["companions"],
                ]:
                    p = root / path
                    p.parent.mkdir(parents=True, exist_ok=True)
                    shutil.copyfile(R / path, p)
                p = root / d["schemas"][0]
                v = g.read(p)
                v["$ref"] = ref
                p.write_text(json.dumps(v))
                with self.assertRaises((AssertionError, KeyError)):
                    g.validate(d, root)


if __name__ == "__main__":
    unittest.main(argv=["interface-test", *rest])

File: tools/shv-interface-codegen/tests/interface_test.cpp
#include "shv_interface.hpp"
#include "test_support.hpp"
using namespace symphony::authoring;
struct InterfaceTests {
  Arguments args;
  std::string module;
  Json original;
  explicit InterfaceTests(const Arguments &selected)
      : args(selected), module("modules/" + args.owner),
        original(read_json(args.root / module / "OWNER-INTERFACE.json")) {}
  void test_frozen_native_parity() {
    const auto hist = read_json(args.root / module /
                                "tests/fixtures/interface-history.v1.json")
                          .at("descriptors");
    test::descriptor(args.engine, shv::latest(hist),
                     original.at("current_version"));
  }
  void test_deterministic_projection_and_drift() {
    shv::validate_v1(original, args.root);
    const auto output = shv::v1_outputs(original);
    test::check(output == shv::v1_outputs(Json(original)),
                "deterministic projections");
    emit(args.root, output, true);
    {
      TempDir temp;
      emit(temp.path, output, false);
      write_bytes(temp.path / output.front().first, "changed");
      test::rejects([&] { emit(temp.path, output, true); }, "generated drift");
    }
  }
  void test_declaration_rejections() {
    const auto old = original.at("releases").begin().key();
    const std::vector<std::pair<std::string, std::function<void(Json &)>>>
        mutations{
            {"extra", [](Json &d) { d["extra"] = true; }},
            {"missing", [](Json &d) { d.erase("schemas"); }},
            {"owner", [](Json &d) { d["module_id"] = "unknown"; }},
            {"namespace", [](Json &d) { d["namespace"] = "foreign"; }},
            {"release",
             [](Json &d) { d["releases"]["latest"] = Json::array(); }},
            {"old admission",
             [&](Json &d) { d["releases"][old].push_back("unknown"); }},
            {"duplicate",
             [](Json &d) {
               const auto op = d["operations"][0];
               d["operations"].push_back(op);
             }},
            {"mutation",
             [](Json &d) { d["operations"][0]["mutability"] = "write"; }},
            {"protocol",
             [](Json &d) {
               d["operations"][0]["output_protocol"] = "invented";
             }},
            {"interaction",
             [](Json &d) {
               d["operations"][0]["administrative_interactions"] =
                   Json::array({"mutate"});
             }},
            {"reader",
             [](Json &d) {
               d["embedded_dependencies"].push_back(
                   {{"engine_id", "foreign"}, {"version", "latest"}});
             }},
            {"escape", [](Json &d) { d["companions"].push_back("../escape"); }},
            {"duplicate schema",
             [](Json &d) {
               const auto schema = d["schemas"][0];
               d["schemas"].push_back(schema);
             }},
            {"missing schema",
             [](Json &d) { d["schemas"].erase(d["schemas"].size() - 1); }}};
    for (const auto &[name, mutate] : mutations) {
      auto bad = original;
      mutate(bad);
      test::rejects([&] { shv::validate_v1(bad, args.root); }, name);
    }
  }
  void test_selected_owner_alias_rejected() {
    TempDir temp;
    const std::string alias = "alias-owner";
    write_json(temp.path / "modules" / alias / "OWNER-INTERFACE.json",
               original);
    test::rejects([&] { shv::load_v1(temp.path, alias); },
                  "frozen owner alias rejected",
                  "selected owner differs from declaration");
    test::equal(shv::load_v1(args.root, args.owner), original,
                "exact selected frozen owner remains accepted");
  }
  void test_duplicate_keys_and_schema_refs() {
    {
      TempDir temp;
      write_bytes(temp.path / "duplicate.json", "{\"a\":1,\"a\":2}");
      test::rejects([&] { read_json(temp.path / "duplicate.json"); },
                    "duplicate keys");
    }
    for (const auto *ref : {"../../outside.json", "#/$defs/Absent"}) {
      TempDir temp;
      test::copy(args.root, temp.path,
                 module + "/tests/fixtures/interface-history.v1.json");
      for (const auto *key : {"schemas", "companions"})
        for (const auto &path : original.at(key))
          test::copy(args.root, temp.path, string(path));
      const auto path = temp.path / string(original.at("schemas").at(0));
      auto doc = read_json(path);
      doc["$ref"] = ref;
      write_json(path, doc);
      test::rejects([&] { shv::validate_v1(original, temp.path); },
                    "schema reference");
    }
  }
};
int main(int argc, char **argv) {
  return main_guard([&] {
    const Arguments args(argc, argv);
    {
      InterfaceTests suite(args);
      std::cout << "RUN test_frozen_native_parity\n";
      suite.test_frozen_native_parity();
      std::cout << "PASS test_frozen_native_parity\n";
    }
    {
      InterfaceTests suite(args);
      std::cout << "RUN test_deterministic_projection_and_drift\n";
      suite.test_deterministic_projection_and_drift();
      std::cout << "PASS test_deterministic_projection_and_drift\n";
    }
    {
      InterfaceTests suite(args);
      std::cout << "RUN test_declaration_rejections\n";
      suite.test_declaration_rejections();
      std::cout << "PASS test_declaration_rejections\n";
    }
    {
      InterfaceTests suite(args);
      std::cout << "RUN test_selected_owner_alias_rejected\n";
      suite.test_selected_owner_alias_rejected();
      std::cout << "PASS test_selected_owner_alias_rejected\n";
    }
    {
      InterfaceTests suite(args);
      std::cout << "RUN test_duplicate_keys_and_schema_refs\n";
      suite.test_duplicate_keys_and_schema_refs();
      std::cout << "PASS test_duplicate_keys_and_schema_refs\n";
    }
    test::done();
  });
}
